### PR TITLE
Fixes #5637 - Don't mark conflicts for confirming ad-hoc leases

### PR DESCRIPTION
**Pull request by GregSutcliffe**
You'll need to reproduce this by adding an active lease without a hostname (ie for a machine that booted up outside of Foreman's control) such that the proxy replies like this:

```
curl -k https://localhost:8443/dhcp/192.168.123.0/52:54:00:33:b5:df
{"starts":"2014-05-09 09:12:22 UTC","ends":"2014-05-09 09:22:22 UTC","state":"active","mac":"52:54:00:33:b5:df","subnet":"192.168.123.0/255.255.255.0","ip":"192.168.123.52"}
```

You can do that with a lease like:

```
lease 192.168.123.52 {
  starts 5 2014/05/09 09:24:23;
  ends 5 2014/05/09 09:34:23;
  cltt 5 2014/05/09 09:24:23;
  binding state active;
  next binding state free;
  rewind binding state free;
  hardware ethernet 52:54:00:33:b5:df;
}
```

Then try to provision a host using that exact mac/ip/network (so the only difference is the missing hostname). The current core code should then crash with a Net::LeaseConflict. This code should provision just fine, but continue to offer conflict resolution when other data is mismatched.

### CODE REVIEW

File: lib/net.rb [hunk 1]
```diff
@@ -11,11 +11,8 @@ def initialize opts = {}
         eval("self.#{k}= v") if self.respond_to?("#{k}=")
       end if opts
 
-      raise Net::LeaseConflict.new("#{self.mac}/#{self.ip}") if opts['state']
-
       self.logger ||= Rails.logger
-      raise "Must define a hostname" if hostname.blank?
```
Comment (domcleal) [inline lib/net.rb:17]: Hostname will still be required for say a DNS record, this validation should probably move.

File: lib/net.rb [hunk 2]
```diff
@@ -11,11 +11,8 @@ def initialize opts = {}
         eval("self.#{k}= v") if self.respond_to?("#{k}=")
       end if opts
 
-      raise Net::LeaseConflict.new("#{self.mac}/#{self.ip}") if opts['state']
```
Comment (domcleal) [inline lib/net.rb:14]: Can the LeaseConflict definition and rescues now be removed from the codebase?

File: lib/net/dhcp/record.rb [hunk 3]
```diff
@@ -49,7 +49,15 @@ def valid?
...
 
     def == other
-      attrs.values_at(:hostname, :mac, :ip, :network) == other.attrs.values_at(:hostname, :mac, :ip, :network)
+      if attrs[:hostname].blank?
+        # If we're converting an 'ad-hoc' lease created by a host booting outside of Foreman's knowledge,
+        # then :hostname will be blank on the incoming lease - if the ip/mac still match, then this
+        # isn't a conflict
+        attrs.values_at(:mac, :ip, :network) == other.attrs.values_at(:mac, :ip, :network)
+      else
+        # Otherwise, if a hostname is present, we want to check all 3 match before declaring a conflict
```
Comment (domcleal) [inline lib/net/dhcp/record.rb]: I count four!

File: lib/net.rb (hunk 1 above)
Comment (GregSutcliffe) [inline lib/net.rb:17]: You can't leave it in though, the incoming lease from record.rb#conflicts doesn't have one

Comment (GregSutcliffe) [inline lib/net.rb:17]: Oh i see, "move". Sorry :)

File: lib/net/dhcp/record.rb (hunk 3 above)
Comment (GregSutcliffe) [inline lib/net/dhcp/record.rb]: I always forget network, it's such a silly check :p

**Comment by domcleal** [comment]
I think we should perhaps have a separate subclass for leases versus records?  It seems records are only meant to model reservations, not leases (see also: sparc_record).

File: lib/net.rb (hunk 2 above)
Comment (GregSutcliffe) [inline lib/net.rb:14]: I would assume so. let me do that.

File: lib/net.rb (hunk 1 above)
Comment (GregSutcliffe) [inline lib/net.rb:17]: Ok done, but untested. Trying it now

**Comment by GregSutcliffe** [comment]
That might make some sense, but would probably involve some proxy-side changes too. Right now when you request /dhcp/:network/:mac the proxy doesn't tell you if its a lease or a reservation - we'd probably want to extend that to report that info back (otherwise we're just guessing from the presence of things like "state").

**Comment by GregSutcliffe** [comment]
[test] that was a weird failure, seeing if it's consistent...

File: lib/net/dns/ptr_record.rb [hunk 4]
```diff
@@ -16,6 +16,7 @@ def destroy
       end
 
       def create
+        raise "Must define a hostname" if hostname.blank?
         super
```
Comment (dmitri-d) [inline lib/net/dns/ptr_record.rb:19]: Shouldn't the same check exist in Net::DNS::ARecord#create ?

Comment (GregSutcliffe) [inline lib/net/dns/ptr_record.rb:19]: Good catch, thanks.

**Comment by dmitri-d** [comment]
lgtm; catches conflicts, skips non-conflicts.

File: lib/net/dhcp/record.rb [hunk 5]
```diff
@@ -49,7 +50,15 @@ def valid?
...
+      if attrs[:hostname].blank?
+        # If we're converting an 'ad-hoc' lease created by a host booting outside of Foreman's knowledge,
+        # then :hostname will be blank on the incoming lease - if the ip/mac still match, then this
+        # isn't a conflict
+        attrs.values_at(:mac, :ip, :network) == other.attrs.values_at(:mac, :ip, :network)
+      else
+        # Otherwise, if a hostname is present, we want to check all 4 match before declaring a conflict
+        attrs.values_at(:hostname, :mac, :ip, :network) == other.attrs.values_at(:hostname, :mac, :ip, :network)
+      end
     end
```
Comment (lzap) [inline lib/net/dhcp/record.rb:62]: One observation - when using dnsmasq, I can see hostnames to be set according to what hosts announce. Isn't possible that ISC DHCP does the same in some configuration? Then `hostname` would not be blank of course.

Good one, thanks for tracking it down.

Comment (GregSutcliffe) [inline lib/net/dhcp/record.rb:62]: This PR doesn't remove conflict resolution entirely - in the situation you describe, the user will still get a "Overwrite" confirmation modal when provisioning the host, and be able to confirm that he wants the record to be written.

Ultimately we can't fix the problem you describe without completely removing conflict detection - having a different hostname _is_ a conflict :)

**Comment by mburns72h** [comment]
I've verified that this works in my environment in combination with the above mentioned smart-proxy patch.

**Comment by lzap** [comment]
I've verified this one too. It works fine now with discovered systems.

**Comment by domcleal** [comment]
Merged as 14646e7 for Foreman 1.5.1, thanks @GregSutcliffe.

**Pull request closed without merging**